### PR TITLE
Precompiled SPIR-V import support

### LINKED ISSUE #4992: SPIR-V library support for import decorations on function declarations
To have linkable spir-v libraries, it will be necessary to allow "import"s.  Any IR lowered to SPIR-V which contains function declarations without definitions should be tagged with import linkage attributes.

### PR DESCRIPTION
**Pull request by cheneym2**
Adds appropriate linkage and function declaration syntax for SPIR-V functions that are declared, to be imported from another SPIR-V module.

Unlike DXIL, stripping the Slang IR for a function down to a declaration requires retaining a block of parameters, as the function declaration must be emitted to SPIR-V with the same parameters as a definition. Because that thwarts the logic in Slang to tell the difference between a declaration and definition, and explicit decoration is introduced to explicitly mark functions which need to be treated as declarations during emit phase.

Fixes #4992

### CODE REVIEW

File: source/slang/slang-ir.cpp [hunk 1]
```diff
@@ -8766,6 +8766,7 @@ namespace Slang
         switch (val->getOp())
         {
         case kIROp_Func:
+            if (val->findDecoration<IRDownstreamModuleImportDecoration>() != nullptr) return false;
```
Comment (csyonghe) [inline source/slang/slang-ir.cpp:8769]: return in its own line.

File: source/slang/slang-ir-redundancy-removal.cpp [hunk 2]
```diff
@@ -177,11 +178,29 @@ void removeAvailableInDownstreamModuleDecorations(CodeGenTarget target, IRModule
                     // Gut the function definition, turning it into a declaration
                     for (auto inst : funcInst->getChildren())
                     {
+                        // For HLSL, we can delete the whole block
```
Comment (csyonghe) [inline source/slang/slang-ir-redundancy-removal.cpp:181]: for (auto block : funcInst->getBlocks())

File: source/slang/slang-ir-redundancy-removal.cpp [hunk 3]
```diff
@@ -177,11 +178,29 @@ void removeAvailableInDownstreamModuleDecorations(CodeGenTarget target, IRModule
...
                         if (inst->getOp() == kIROp_Block)
                         {
-                            toRemove.add(inst);
+                            if (target == CodeGenTarget::HLSL)
+                            {
+                                toRemove.add(inst);
+                            }
+                            else if (target == CodeGenTarget::SPIRV)
+                            {
+                                for (auto blockInst : inst->getChildren())
```
Comment (csyonghe) [inline source/slang/slang-ir-redundancy-removal.cpp:192]: You cannot modify the children while iterating it with getChildren.
need to call `inst->getModifiableChildren()`.

See https://github.com/shader-slang/slang/blob/master/docs/design/ir.md#ir-global-and-hoistable-value-deduplication

File: source/slang/slang-ir-redundancy-removal.cpp [hunk 4]
```diff
@@ -177,11 +178,29 @@ void removeAvailableInDownstreamModuleDecorations(CodeGenTarget target, IRModule
                     // Gut the function definition, turning it into a declaration
                     for (auto inst : funcInst->getChildren())
                     {
+                        // For HLSL, we can delete the whole block
+                        // but for SPIR-V we have to keep the block with only the
+                        // parameters inside, no other instructions.
                         if (inst->getOp() == kIROp_Block)
                         {
-                            toRemove.add(inst);
+                            if (target == CodeGenTarget::HLSL)
```
Comment (csyonghe) [inline source/slang/slang-ir-redundancy-removal.cpp:186]: Why can't HLSL be handled by the same SPIRV path? I would think that HLSL emitter also needs to know the parameters to be able to emit the correct declaration.

File: source/slang/slang-ir-redundancy-removal.cpp (hunk 3 above)
Comment (cheneym2) [reply source/slang/slang-ir-redundancy-removal.cpp:192]: Isn't it acceptable that I'm just adding the block instructions to a list of things, and then I remove them all at the end?

File: source/slang/slang-ir-redundancy-removal.cpp (hunk 4 above)
Comment (cheneym2) [reply source/slang/slang-ir-redundancy-removal.cpp]: That hasn't been an issue so far, so it's confusing to me.  HLSL emission manages to preserve the function parameters even when deleting all function blocks.  I don't understand how it's able to.  The code that recognized empty functions and emitted HLSL function declarations existed already and I have yet to see how it worked.

Current code still produces desirable declarations for HLSL:

```
int normalFunc_0(int _0);
int normalFuncUsesGeneric_0(int _0);
```

File: source/slang/slang-ir-redundancy-removal.cpp (hunk 3 above)
Comment (csyonghe) [reply source/slang/slang-ir-redundancy-removal.cpp]: Ah you are right that works too.

File: source/slang/slang-ir-redundancy-removal.cpp (hunk 4 above)
Comment (csyonghe) [reply source/slang/slang-ir-redundancy-removal.cpp]: Can we unify the two cases a bit by having spirv emit logic use the func type to emit params too?

We want to have some consistency in our IR in terms of what it means for all backends.

Comment (cheneym2) [reply source/slang/slang-ir-redundancy-removal.cpp]: It should be doable.  I see emitFuncDecl (c-like) as the example of what HLSL does.
I'm confused why function declarations in SPIR-V require OpFunctionParameters.  It seems redundant to the OpTypeFunction which is already a field of the OpFunction itself.

Comment (csyonghe) [reply source/slang/slang-ir-redundancy-removal.cpp]: That's probably to allow you to put decorations on the parameters.

File: source/slang/slang-emit-spirv.cpp [hunk 5]
```diff
@@ -2631,14 +2632,75 @@ struct SPIRVEmitContext
...
         //
-        m_sink->diagnose(irFunc, Diagnostics::internalCompilerError);
-        SLANG_UNEXPECTED("function declaration in SPIR-V emit");
-        UNREACHABLE_RETURN(nullptr);
+        auto funcType = irFunc->getDataType();
+        auto paramCount = funcType->getParamCount();
+        for (UInt pp = 0; pp < paramCount; ++pp)
+        {
+            auto paramType = funcType->getParamType(pp);
+            emitOpFunctionParameter(spvFunc, nullptr, paramType);
```
Comment (csyonghe) [inline source/slang/slang-emit-spirv.cpp:2684]: Need to make sure to emit necessary decorations for pointer-typed parameters as well. Need to call `maybeEmitPointerDecoration` here.

Comment (cheneym2) [reply source/slang/slang-emit-spirv.cpp:2684]: Looking at the implementation, it's for PhysicalStorageBuffer pointers and I'm unclear how to write a test case for that.  What function parameter would I define in Slang that corresponds to that in spir-v?
I'm currently trying to find a test case that hits this test case for function definitions to model after.

Comment (cheneym2) [reply source/slang/slang-emit-spirv.cpp:2684]: Nm, I believe I have a test case now.

Comment (cheneym2) [reply source/slang/slang-emit-spirv.cpp:2684]: Exporting this shows an "Alias" decoration applied from the function definition side.
```
public int normalFunc(int a, const float *b)
{
    return a - floor(*b);
}
```

Trying to replicate that on the import side function declaration.

File: source/slang/slang-emit-spirv.cpp [hunk 6]
```diff
@@ -2628,14 +2630,76 @@ struct SPIRVEmitContext : public SourceEmitterBase, public SPIRVEmitSharedContex
...
+        // blocks in a function declaration, so we will emit the parameters
+        // as derived from the function's type.
+        //
+        auto funcType = irFunc->getDataType();
+        auto paramCount = funcType->getParamCount();
+        for (UInt pp = 0; pp < paramCount; ++pp)
+        {
+            auto paramType = funcType->getParamType(pp);
+            SpvInst* spvParam = emitOpFunctionParameter(spvFunc, nullptr, paramType);
+            maybeEmitPointerDecoration(spvParam, paramType, false, kIROp_Param);
```
Comment (cheneym2) [inline source/slang/slang-emit-spirv.cpp:2683]: Added pointer decoration

**Review by csyonghe — DISMISSED** [review]

**Review by csyonghe — APPROVED** [review]

**Pull request merged**